### PR TITLE
Add support for Concat Operation

### PR DESCRIPTION
**Pull request by ramakota-helprack**
Added lit_tests for concat op.

### CODE REVIEW

File: src/Conversion/ONNXToTorch/NN/ConcatOp.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    ValueRange inputs = op1.inputs();
+    auto axisValue = op1.axisAttr();       // ::mlir::IntegerAttr
+    Value  axisVal = rewriter.create<ConstantIntOp>(loc, axisValue);
+    
+    auto resultType = toTorchType(context, op1.getType());
+    std::vector<Value> inputArrayValues;
+
+    // Iterate through the list of inputs and create corresponding
+    // torch tensors for each input tensor.
+    for (Value input : inputs)
```
Comment (BraunPhilipp) [inline src/Conversion/ONNXToTorch/NN/ConcatOp.cpp:78]: Cleaned up your code a bit. Type will generally be the same. Input and result type will remain the same.

**Review by BraunPhilipp — APPROVED** [review]

**Comment by BraunPhilipp** [comment]
Otherwise, LGTM.

File: src/Conversion/ONNXToTorch/NN/ConcatOp.cpp (hunk 1 above)
Comment (ramakota-helprack) [reply src/Conversion/ONNXToTorch/NN/ConcatOp.cpp:78]: @BraunPhilipp, i am getting inputType and resultType are different. 
E.g: 
%0 = "onnx.Concat"(%arg0, %arg0) {axis = 1 : si64, onnx_node_name = "Concat_0"} : (tensor<5x5xf32>, tensor<5x5xf32>) -> tensor<5x10xf32>

from the above concat stmt, each inputType is "tensor<5x5xf32>" and resultType is " tensor<5x10xf32>". Both are different. 

I think your comment is related these inputType and resultType only right ? if not please correct me.

Comment (BraunPhilipp) [reply src/Conversion/ONNXToTorch/NN/ConcatOp.cpp:78]: This was related to the element type and not the shape.

**Review by BraunPhilipp — APPROVED** [review]

File: test/mlir/conversion/onnx_to_torch/concat/concat_axis_three.onnx.mlir [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
...D1606653D566BD2BC36A7FDBD6B4A173EC7F7253E069D31BDA43F0A3EEB3F98BCA0C3BDBC33EE90BD20CCA8BC80890D3EA68A35BDACED0EBE680EEABD66312C3D589A973D1050003E3647E33CFB74DABDC380A13D80D8F83D001B48BD00C09F3B07F5213E68E7B73DE4F11F3EBB641DBE8B451EBE88538ABDB00605BED6C2283C230683BD36F702BE20383FBDE6E82FBD160CE1BD78040F3EC09C1F3C8B9CAABC53DDCCBDA8A4ED3DDE6806BEE6A739BD563D0DBD8B6079BDC06BFC3CF37BB23DFEB1D6BD58C790BD4870BF3DAE40FC3DF66FE13C5623C03C0B66CCBD56178B3AD677B53C06326D3D84B6053ECE08A8BD033E8FBDA83B833D4016CFBC70731CBE84890C3E36278C3D5877833DFBA9573DD6FEF83C205EBD3CAB3B25BEC68318BEBEC4E83D801EDFBC6352063E5809F3BDDB7F00BD38AF873D0E0E1CBE4B80023D9331D2BDEB54E6BC57DA0ABE137FCFBD1024223D5BE812BEF89021BE58A4B63D1E1CDABDBBF91E3E5B5C46BD80E78EBB3677193E239A1ABE83D1913D673E17BEEBF7E1BCC8799ABD5690F8BD2849F2BD84E11B3ED8FCC4BD8BC1EABCAB03653B8882A33D7659BCBDAB48CD3CC0A571BDAF4407BEF7A80ABE6B01163EAB79F73C13EEEA3DC36726BE93FA8FBDC6404DBDDF841E3ECFE725BE6BA771BC1B5A23BECAAF043E6012ABBD10D2EABD3A62043E2041333DF3F899BDFB03F3BD77FC21BE5824B0BD9B3E553D8C7A0ABEC628333DC8DACE3D3B69D2BDFE2F92BD004069BC98B11A3E12AF0FBE3EBEED3D1608ACBDDFA813BE503DA5BDC305F9BD"> : tensor<12x4x3x3xf32>} : () -> tensor<12x4x3x3xf32>
```
Comment (BraunPhilipp) [inline test/mlir/conversion/onnx_to_torch/concat/concat_axis_three.onnx.mlir:4]: nit: We can probably remove the weights here.

Comment (ramakota-helprack) [reply test/mlir/conversion/onnx_to_torch/concat/concat_axis_three.onnx.mlir:4]: @anushree1998, Please check once.

Comment (anushree1998) [reply test/mlir/conversion/onnx_to_torch/concat/concat_axis_three.onnx.mlir:4]: We have removed the weights.

File: test/mlir/conversion/onnx_to_torch/Concat/concat_axis_three.onnx.mlir [hunk 3]
```diff
@@ -0,0 +1,13 @@
+//RUN: onnx-mlir --EmitONNXIR --run-torch-pass %s -o=%t >/dev/null && cat %t.onnx.mlir | FileCheck -v %s
+module attributes {}  {
+  func @main_graph(%arg0: tensor<1x4x12x12xf32>) -> tensor<1x12x12x24xf32> attributes {input_names = ["input"], output_names = ["4"]} {
+    %0 = "onnx.Constant"() {value = dense<"0xDEADBEEF"> : tensor<12x4x3x3xf32>} : () -> tensor<12x4x3x3xf32>
+    %1 = "onnx.Constant"() {value = dense<[-0.0787997097, 0.0934591516, -0.106071398, 0.105203271, 0.0158753395, -0.0936289653, -0.15529643, 0.165667564, 0.0695708394, -0.0372162871, 0.166498899, -0.0204498172]> : tensor<12xf32>} : () -> tensor<12xf32>
```
Comment (BraunPhilipp) [inline test/mlir/conversion/onnx_to_torch/Concat/concat_axis_three.onnx.mlir:5]: nit: Please remove weights consistently. Also keep mlir folder names lowercase or uppercase.

**Review by BraunPhilipp — APPROVED** [review]

File: test/mlir/conversion/onnx_to_torch/Concat/concat_axis_three.onnx.mlir (hunk 3 above)
Comment (anushree1998) [reply test/mlir/conversion/onnx_to_torch/Concat/concat_axis_three.onnx.mlir:5]: Hi Braun,
We have removed weights and changed mlir folder names to lowercase also.

**Review by siboehm — CHANGES_REQUESTED** [review]
Tests zero&one, and two&three are very similar, so one of each should suffice. Both also only concatenate identical tensors (`onnx.Concat(%2, %2)` and `onnx.Concat(%arg0, %arg0)`). 

Can we add a test that say, concatenates 5 tensors, each of a different shape?

**Comment by anushree1998** [comment]
Hi siboehm,
We are looking at this.

**Comment by anushree1998** [comment]
Hi @siboehm ,
We have added two more test cases where concatenation is happening with different shaped tensors.

**Review by siboehm — APPROVED** [review]
Thanks for the changes, tests are looking more thorough now!

**Pull request merged**